### PR TITLE
PYTHON-2680 Breaking changes to DBRef BSON+JSON decoding

### PR DESCRIPTION
**Pull request by ShaneHarvey**
Still TODO:

- [x] Implement DBRef spec tests (https://github.com/mongodb/specifications/blob/master/source/dbref.rst#test-plan)
- [x] Add changelog and migration docs.

### CODE REVIEW

File: bson/__init__.py [hunk 1]
```diff
@@ -202,7 +202,10 @@ def _get_object(data, view, position, obj_end, opts, dummy):
     obj = _elements_to_dict(data, view, position + 4, end, opts)
 
     position += obj_size
-    if "$ref" in obj:
+    # If DBRef validation fails, return a normal doc.
```
Comment (ShaneHarvey) [inline bson/__init__.py:205]: This is the breaking DBRef BSON decoding change.

File: bson/json_util.py [hunk 2]
```diff
@@ -463,7 +460,9 @@ def object_pairs_hook(pairs, json_options=DEFAULT_JSON_OPTIONS):
 def object_hook(dct, json_options=DEFAULT_JSON_OPTIONS):
     if "$oid" in dct:
         return _parse_canonical_oid(dct)
-    if "$ref" in dct:
+    if (isinstance(dct.get('$ref'), str) and
+            "$id" in dct and
+            isinstance(dct.get('$db'), (str, type(None)))):
```
Comment (ShaneHarvey) [inline bson/json_util.py:465]: This is the breaking DBRef JSON decoding change.

File: bson/json_util.py [hunk 3]
```diff
@@ -666,15 +665,16 @@ def _parse_canonical_regex(doc):
     if len(regex) != 2:
         raise TypeError('Bad $regularExpression must include only "pattern"'
                         'and "options" components: %s' % (doc,))
-    return Regex(regex['pattern'], regex['options'])
+
+    # Validate options, defer to Regex for pattern validation.
+    options = regex['options']
+    if not isinstance(options, str):
+        raise TypeError("pattern must be a string, not %s" % type(options))
+    return Regex(regex['pattern'], options)
```
Comment (ShaneHarvey) [inline bson/json_util.py:673]: Ignore this, I'll rebase on top of https://github.com/mongodb/mongo-python-driver/pull/721

**Review by behackett — APPROVED** [review]

**Comment by ShaneHarvey** [comment]
@juliusgeo I've merged this without waiting for your review but please take a look through this change anyway. Feel free to ask any questions you might have.

**Pull request merged**